### PR TITLE
Fix single core coverity violations in FreeRTOS-Kernel.

### PR DESCRIPTION
**Pull request by ActoryOu**
Fix single core coverity violations in FreeRTOS-Kernel.

Description
-----------

Fix single core coverity violations in FreeRTOS-Kernel.

Test Steps
-----------

1. Clone my forked [FreeRTOS repo](https://github.com/ActoryOu/FreeRTOS/tree/kernel-coverity).
1. cd FreeRTOS/Test/Coverity
1. Update the compiler configuration in Coverity
`cov-configure --force --compiler cc --comptype gcc`
`cov-configure -co /usr/bin/cc -- -m32 -std=gnu90`
1. Build the (pseudo) application
`cov-build --emit-complementary-info --dir cov-out make singlecore`
1. Go to the Coverity output directory (cov-out) and begin Coverity static analysis
`cd cov-out/`
`cov-analyze --dir . --coding-standard-config ../../FreeRTOS/Test/Coverity/coverity_misra.config`
1. Format the errors in HTML format so that it is more readable while removing the FreeRTOS-Kernel directory from the report
`cov-format-errors --dir . --exclude-files '(/Test/)' --html-output ./output`

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: event_groups.c [hunk 1]
```diff
@@ -120,10 +128,7 @@ static BaseType_t prvTestWaitCondition( const EventBits_t uxCurrentEventBits,
 
             #if ( configSUPPORT_DYNAMIC_ALLOCATION == 1 )
             {
-                /* Both static and dynamic allocation can be used, so note that
-                 * this event group was created statically in case the event group
-                 * is later deleted. */
-                pxEventBits->ucStaticallyAllocated = pdTRUE;
+                pxEventBits->ucStaticallyAllocated = ( uint8_t ) pdTRUE;
```
Comment (chinglee-iot) [inline event_groups.c]: Suggest to keep the original comment.

File: event_groups.c [hunk 2]
```diff
@@ -199,13 +204,15 @@ EventBits_t xEventGroupSync( EventGroupHandle_t xEventGroup,
     EventBits_t uxOriginalBitValue, uxReturn;
     EventGroup_t * pxEventBits = xEventGroup;
     BaseType_t xAlreadyYielded;
-    BaseType_t xTimeoutOccurred = pdFALSE;
+    BaseType_t xTimeoutOccurred = ( BaseType_t ) pdFALSE;
+    /* Declare to confort MISRA C 2012 Rule 17.8 - "A function parameter should not be modified" */
```
Comment (chinglee-iot) [inline event_groups.c]: Suggest to remove the comment.
```suggestion

```

File: event_groups.c [hunk 3]
```diff
@@ -331,7 +339,9 @@ EventBits_t xEventGroupWaitBits( EventGroupHandle_t xEventGroup,
     EventGroup_t * pxEventBits = xEventGroup;
     EventBits_t uxReturn, uxControlBits = 0;
     BaseType_t xWaitConditionMet, xAlreadyYielded;
-    BaseType_t xTimeoutOccurred = pdFALSE;
+    BaseType_t xTimeoutOccurred = ( BaseType_t ) pdFALSE;
+    /* Declare to confort MISRA C 2012 Rule 17.8 - "A function parameter should not be modified" */
```
Comment (chinglee-iot) [inline event_groups.c]: Suggest to remove the comment.
```suggestion

```

File: event_groups.c [hunk 4]
```diff
@@ -368,20 +378,21 @@ EventBits_t xEventGroupWaitBits( EventGroupHandle_t xEventGroup,
...
             }
         }
-        else if( xTicksToWait == ( TickType_t ) 0 )
+        else if( xLocalTicksToWait == ( TickType_t ) 0 )
         {
             /* The wait condition has not been met, but no block time was
              * specified, so just return the current value. */
             uxReturn = uxCurrentEventBits;
-            xTimeoutOccurred = pdTRUE;
+
```
Comment (chinglee-iot) [inline event_groups.c]: Suggest to remove the extra space.
```suggestion

```

File: include/FreeRTOS.h [hunk 5]
```diff
@@ -768,7 +768,7 @@
 #endif
 
 #ifndef traceTASK_INCREMENT_TICK
-    #define traceTASK_INCREMENT_TICK( xTickCount )
+    #define traceTASK_INCREMENT_TICK( xTickToIncrement )
```
Comment (chinglee-iot) [inline include/FreeRTOS.h]: This looks like a typo?

File: timers.c [hunk 6]
```diff
@@ -254,7 +258,7 @@
                                                       configTIMER_SERVICE_TASK_NAME,
                                                       ulTimerTaskStackSize,
                                                       NULL,
-                                                      ( ( UBaseType_t ) configTIMER_TASK_PRIORITY ) | portPRIVILEGE_BIT,
+                                                      ( UBaseType_t ) ( ( uint8_t ) configTIMER_TASK_PRIORITY ) | portPRIVILEGE_BIT,
```
Comment (chinglee-iot) [inline timers.c]: It looks like we cast to ( uint8_t ) then cast to ( UBaseType_t ). Why are we doing this?

**Comment by chinglee-iot** [comment]
There are some changes. This PR should target coverity violation only. Suggest we change the logic in other PR.
```
!= pdFALSE
=>
== pdTRUE
```

File: tasks.c [hunk 7]
```diff
@@ -321,7 +350,7 @@ typedef struct tskTaskControlBlock       /* The old naming convention is used to
 
     #if ( configUSE_TRACE_FACILITY == 1 )
         UBaseType_t uxTCBNumber;  /*< Stores a number that increments each time a TCB is created.  It allows debuggers to determine when a task has been deleted and then recreated. */
-        UBaseType_t uxTaskNumber; /*< Stores a number specifically for use by third party trace code. */
+        UBaseType_t uxTaskID; /*< Stores a number specifically for use by third party trace code. */
```
Comment (chinglee-iot) [inline tasks.c]: Suggest we change the name of static variable, uxTaskNumber, instead of this member variable name of TCB.

File: tasks.c [hunk 8]
```diff
@@ -363,18 +397,27 @@ typedef struct tskTaskControlBlock       /* The old naming convention is used to
...
 
 /* The old tskTCB name is maintained above then typedefed to the new TCB_t name
  * below to enable the use of older kernel aware debuggers. */
+typedef struct tskTaskControlBlock tskTCB;
 typedef tskTCB TCB_t;
 
 /*lint -save -e956 A manual analysis and inspection has been used to determine
  * which static variables must be declared volatile. */
 #if ( configNUM_CORES == 1 )
+    /* Declare an extern here to confort MISRA C 2012 Rule 8.4 -
```
Comment (chinglee-iot) [inline tasks.c]: Suggest to use suppress instead of extern.

File: tasks.c [hunk 9]
```diff
@@ -363,18 +397,27 @@ typedef struct tskTaskControlBlock       /* The old naming convention is used to
...
+     * "A compatible declaration shall be visible when an object or function with external linkage is defined"
+     * This variable is used in some ports. */
+    extern portDONT_DISCARD PRIVILEGED_DATA TCB_t * volatile pxCurrentTCB;
     portDONT_DISCARD PRIVILEGED_DATA TCB_t * volatile pxCurrentTCB = NULL;
 #else
-portDONT_DISCARD PRIVILEGED_DATA TCB_t * volatile pxCurrentTCBs[ configNUM_CORES ] = { NULL };
+    /* Declare an extern here to confort MISRA C 2012 Rule 8.4 - 
+     * "A compatible declaration shall be visible when an object or function with external linkage is defined"
+     * This variable is used in some ports. */
+    extern portDONT_DISCARD PRIVILEGED_DATA TCB_t * volatile pxCurrentTCBs[ configNUM_CORES ];
```
Comment (chinglee-iot) [inline tasks.c]: Suggest to use suppress instead of extern.

File: tasks.c [hunk 10]
```diff
@@ -423,6 +466,10 @@ PRIVILEGED_DATA static TaskHandle_t xIdleTaskHandles[ configNUM_CORES ] = { NULL
 /* Improve support for OpenOCD. The kernel tracks Ready tasks via priority lists.
  * For tracking the state of remote threads, OpenOCD uses uxTopUsedPriority
  * to determine the number of priority lists to read back from the remote target. */
+/* Declare an extern here to confort MISRA C 2012 Rule 8.4 - 
+ * "A compatible declaration shall be visible when an object or function with external linkage is defined"
+ * This variable is used in OpenOCD. */
+extern const volatile UBaseType_t uxTopUsedPriority;
```
Comment (chinglee-iot) [inline tasks.c]: Suggest to use suppress instead of extern.

File: tasks.c [hunk 11]
```diff
@@ -1437,27 +1490,29 @@ static void prvInitialiseNewTask( TaskFunction_t pxTaskCode,
...
-        uxPriority &= ~portPRIVILEGE_BIT;
+        uxLocalPriority &= ~portPRIVILEGE_BIT;
     #endif /* portUSING_MPU_WRAPPERS == 1 */
 
     /* Avoid dependency on memset() if it is not required. */
     #if ( tskSET_NEW_STACKS_TO_KNOWN_VALUE == 1 )
     {
         /* Fill the stack with a known value to assist debugging. */
-        ( void ) memset( pxNewTCB->pxStack, ( int ) tskSTACK_FILL_BYTE, ( size_t ) ulStackDepth * sizeof( StackType_t ) );
+        ( void ) memset( pxNewTCB->pxStack, ( BaseType_t ) tskSTACK_FILL_BYTE, ( UBaseType_t ) ulStackDepth * sizeof( StackType_t ) );
```
Comment (chinglee-iot) [inline tasks.c]: Th prototype of 
```
memset (void *dstpp, int c, size_t len)
```
Are we using our own implementation?

File: tasks.c [hunk 12]
```diff
@@ -3143,6 +3208,15 @@ static BaseType_t prvCreateIdleTasks( void )
             #endif /* #if ( configNUM_CORES > 1 ) */
         }
         #endif /* configSUPPORT_STATIC_ALLOCATION */
+
```
Comment (chinglee-iot) [inline tasks.c:3182]: This looks like a bug fix. Can we put it in another PR?

File: include/FreeRTOS.h (hunk 5 above)
Comment (ActoryOu) [reply include/FreeRTOS.h:771]: xTickCount is actually a global variable in tasks.c, MISRA doesn't allow this kind of condition.

File: tasks.c [hunk 13]
```diff
@@ -3380,7 +3454,13 @@ void vTaskSuspendAll( void )
         }
         else
         {
-            xReturn = xNextTaskUnblockTime - xTickCount;
+            /* Declare xLocalNextTaskUnblockTime/xLocalTickCount here to confort MISRA C 2012 Rule 13.2 -
```
Comment (chinglee-iot) [inline tasks.c]: Suggest we split it into the following instruction instead of using local variables
```
xReturn = xNextTaskUnblockTime;
xReturn = xReturn - xTickCount;
```

File: tasks.c [hunk 14]
```diff
@@ -5271,7 +5373,11 @@ static void prvCheckTasksWaitingTermination( void )
                 #if ( configNUM_CORES == 1 )
                 {
                     pxTCB = listGET_OWNER_OF_HEAD_ENTRY( ( &xTasksWaitingTermination ) ); /*lint !e9079 void * is used as this macro is used with timers and co-routines too.  Alignment is known to be fine as the type of the pointer stored and retrieved is the same. */
-                    ( void ) uxListRemove( &( pxTCB->xStateListItem ) );
+
+                    if( pxTCB != NULL )
```
Comment (chinglee-iot) [inline tasks.c]: Suggest to use configASSERT here.

File: tasks.c [hunk 15]
```diff
@@ -5284,19 +5390,22 @@ static void prvCheckTasksWaitingTermination( void )
                     {
                         pxTCB = listGET_OWNER_OF_HEAD_ENTRY( ( &xTasksWaitingTermination ) ); /*lint !e9079 void * is used as this macro is used with timers and co-routines too.  Alignment is known to be fine as the type of the pointer stored and retrieved is the same. */
 
-                        if( pxTCB->xTaskRunState == taskTASK_NOT_RUNNING )
+                        if( pxTCB != NULL )
```
Comment (chinglee-iot) [inline tasks.c]: Suggest to use configASSERT here.

**Review by chinglee-iot — COMMENTED** [review]
Also suggest we put 
MISRA 15.5 advisory  "A function should have a single point of exit at the end" to another PR.

File: include/FreeRTOS.h (hunk 5 above)
Comment (ActoryOu) [reply include/FreeRTOS.h]: Reverted due to suppress all advisory checkers.

File: tasks.c (hunk 13 above)
Comment (ActoryOu) [reply tasks.c]: Suppress because it's scanned by advisory checker.

File: tasks.c (hunk 12 above)
Comment (ActoryOu) [reply tasks.c:3182]: It's not a bug fix. Move this section just to confort MISRA checker for single core, but the logic is same.
